Fragments of Data Cleaning Chapter

tillahoffmann opened the pull request:
This PR adds the first few sections of the Data Cleaning chapter; it replaces #7. Specifically,

- `01-intro.md` containing a brief overview and learning objectives.
- `02-reshaping-data.md` discussing long and wide data and transposition.
- `03-tidying-data.md` discussing row and column filtering, summary statistics, grouping, sorting, etc.
- `Makefile` that builds html versions of the markdown documents including execution. We won't need this to present the course, but it's a nice sanity check that the code in the notebooks actually runs.
- `requirements.in` and `requirements.txt` which specified the versions of various python packages to avoid runtime environment drift.
- `.github/workflows/python-markdown.yml` which sets up a python environment and executes the notebooks in CI. I may need a few iterations to get it to build; sorry about the inbox noise.

A few notes:

- I've stuck the files in a folder for the chapter to keep things organized and used `.md` extensions because the content isn't R. I don't know if that messes with the ottr build?
- Thought I'd get your input on the first few sections of the chapter before committing to the remaining sections. All input welcome!
- I've omitted the "What is Tidy Data?" section because it's not Python-specific and already well covered in the existing course.

tillahoffmann committed:
Add first two sections of data cleaning chapter.

tillahoffmann committed:
Move chapters to root directory.

tillahoffmann committed:
Add section indicator.

tillahoffmann committed:
Add output as markdown tables to reshaping section.

tillahoffmann commented:
Converted to draft for now. Will rebase and reopen once #13 is merged.

cansavvy reviewed (commented):
Once you get the branches sorted out, I think the content that is here so far looks great! Very excited for this. Only some minor comments, but the general strategies and content look great to me!

tillahoffmann commented on 05-cleaning-data_02-reshaping-data.md:
```diff
@@ -0,0 +1,154 @@
+
+Converting your data from wide-to-long or from long-to-wide data formats is referred to as **reshaping** your data. You can do this with the [`pandas``](https://pandas.pydata.org) package. In fact, we will use the pandas package for all of our data tidying, wrangling, and reshaping. As with most helpful Python packages, there is more functionality than what is discussed here, so feel free to expl ...
+
+For these examples, we'll work with the [`relig_income`](https://tidyr.tidyverse.org/reference/relig_income.html)  dataset available in `tidyr`. The data in this dataset includes the income ranges and religions from the Pew religion and income survey. The following code downloads the data from GitHub and displays the first few rows using the [`head`](https://pandas.pydata.org/docs/reference/api/p ...
```
Good point. I took a somewhat content-blind approach translating the existing R resources to Python. But maybe this is a good opportunity to review some of the material, too. I'll have a look for other datasets.

tillahoffmann replied:
I've been using the `palmerpenguins` dataset (https://allisonhorst.github.io/palmerpenguins/) for examples. It's quite neat and could probably illustrate the same concepts.

cansavvy replied:
Penguins are lovely! And that's a nice dataset to work with. Definitely a classic.

tillahoffmann replied:
I've been pondering on this a bit more, and the penguin dataset is already "tidy" so a number of sections in the chapter aren't that relevant. Do you know of another dataset that has untidy structure but is maybe a less sensitive dataset?

tillahoffmann replied:
The `billboard` charts from 2000 seem to do the trick.

tillahoffmann commented on 05-cleaning-data_02-reshaping-data.md:
```diff
@@ -0,0 +1,154 @@
+
+## Additional Resources
+
+* [TidyDataTutor](https://tidydatatutor.com/vis.html).
```
```suggestion
* [PandasTidyDataTutor](https://pandastutor.com/)
```

tillahoffmann reviewed (commented):
Thank you for the review! A few follow-up questions inline.

tillahoffmann committed:
Reinstate file references in `_bookdown.yml` lost during rebase.

tillahoffmann committed:
Add reference to pandas tutor and fix typo.

cansavvy commented on 05-cleaning-data_02-reshaping-data.md:
```diff
@@ -0,0 +1,154 @@
+
+Long data, on the other hand, has one column indicating the type of variable contained in that row and then a separate row for the value for that variable. Each row contains a single observation for a single variable.  Below is an example of a long data set.
+
+| ID | Variable | Value |
```
The docker image OTTR uses has python on it so we could easily add to the github action to rerun the python code and then you could have it so the OTTR markdown just displays that through an iframe or something by knitting in a smaller markdown.

But this is probably not 100% necessary unless we forsee a change in the table that would change the interpretation of the chapter.

tillahoffmann replied:
I'm prone to falling into the [automation trap](https://xkcd.com/1319/) so stuck with the copy-paste for now. Definitely up for revisiting if we end up wanting to include a lot of outputs.

tillahoffmann committed:
Use `billboard` instead of `relig_income` dataset.

tillahoffmann committed:
Fix broken example url.

tillahoffmann committed:
Ignore playground folder.

tillahoffmann committed:
Fix typos and update dictionary.

tillahoffmann committed:
Further dictionary exceptions.

tillahoffmann commented:
Thanks for the review, @davonperson! 

I have updated the example dataset to use billboard charts rather than the survey on religion and income. I had to make a few changes to the dictionary and renamed `LICENSE.md` to `LICENSE.txt` so it wouldn't get picked up by the spellchecker. I think this is now ready to be merged.
